File: apps/api/src/domain/entities/task.py
"""
Domain Entity: Task
Represents a task in the system following DDD principles
"""

from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID, uuid4


class TaskStatus(str, Enum):
    """Task status enumeration"""

    PENDING = "PENDING"
    IN_PROGRESS = "IN_PROGRESS"
    COMPLETED = "COMPLETED"
    CANCELLED = "CANCELLED"


@dataclass
class TaskId:
    value: UUID


@dataclass
class Task:
    """Task domain entity"""

    id: TaskId
    title: str
    description: str
    status: TaskStatus
    assignee_id: Optional[UUID]
    created_at: datetime
    updated_at: datetime

    @classmethod
    def create(cls, title: str, description: str, assignee_id: Optional[UUID] = None) -> "Task":
        """Factory method to create a new task"""
        now = datetime.utcnow()
        return cls(
            id=TaskId(uuid4()),
            title=title,
            description=description,
            status=TaskStatus.PENDING,
            assignee_id=assignee_id,
            created_at=now,
            updated_at=now,
        )

    def update_status(self, status: TaskStatus) -> None:
        """Update task status"""
        self.status = status
        self.updated_at = datetime.utcnow()

    def assign_to(self, user_id: UUID) -> None:
        """Assign task to a user"""
        self.assignee_id = user_id
        self.updated_at = datetime.utcnow()

File: apps/web/src/domain/entities/Task.ts
// Domain Entity: Task
// Represents a task in the system following DDD principles

export interface TaskId {
  value: string
}

export enum TaskStatus {
  PENDING = 'PENDING',
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
  CANCELLED = 'CANCELLED',
}

export interface TaskProps {
  id: TaskId
  title: string
  description: string
  status: TaskStatus
  assigneeId?: string
  createdAt: Date
  updatedAt: Date
}

export class Task {
  private constructor(private props: TaskProps) {}

  static create(
    props: Omit<TaskProps, 'createdAt' | 'updatedAt' | 'status'>
  ): Task {
    return new Task({
      ...props,
      status: TaskStatus.PENDING,
      createdAt: new Date(),
      updatedAt: new Date(),
    })
  }

  static fromPersistence(props: TaskProps): Task {
    return new Task(props)
  }

  get id(): TaskId {
    return this.props.id
  }

  get title(): string {
    return this.props.title
  }

  get description(): string {
    return this.props.description
  }

  get status(): TaskStatus {
    return this.props.status
  }

  get assigneeId(): string | undefined {
    return this.props.assigneeId
  }

  get createdAt(): Date {
    return this.props.createdAt
  }

  get updatedAt(): Date {
    return this.props.updatedAt
  }

  updateStatus(status: TaskStatus): void {
    this.props.status = status
    this.props.updatedAt = new Date()
  }

  assignTo(userId: string): void {
    this.props.assigneeId = userId
    this.props.updatedAt = new Date()
  }

  toPersistence(): TaskProps {
    return { ...this.props }
  }
}
